### DISCUSSION TITLE
New moon's crescent visibility map

### DISCUSSION THREAD

**Discussion by ebraminio** [skyfielders/python-skyfield#743] [Q&A]
Hi! I am trying to create a kind of map which have puzzled me for days which at the end should look like this:

<img width="450" alt="image" src="https://user-images.githubusercontent.com/833473/167357775-1b59ca98-f462-4bb3-a75d-69b0a316af9d.png">

[source](https://www.moonsighting.com/visibility.html)

From [this paper](https://web.archive.org/web/20060628073223/http://web.lums.edu.pk:80/~casm/winterconference-fullpapers/shahidqureshi-paper.pdf ), page 13 and 14, I learnt one formulation to create such maps is called Yallop which here is [its paper](https://www.oocities.org/hazarry/astronomy/yallop_1997.pdf) (and from [this code](https://github.com/AminKH/MoonSun/blob/19d2e5e3fe2bee657ce24e306162f55e6989687e/Fortran/MoonSun.f95#L3293-L3297) I learnt there are other formulations also such as Odeh, [its paper](https://bruneiastronomy.org/web/wp-content/uploads/2015/07/New_Criterion_Crescent.pdf), which should give the same but different intensity)

I searched the term Yallop in GitHub and found [this implementation](https://github.com/aloistr/swisseph/blob/d9476bf5d6aefd85f07c79ccb3c0df8fd0a3e706/swehel.c#L1934) and in combination of [this code from Swiss Ephemeris](https://github.com/msyazwanfaid/Authentication-Algorithm-for-New-Moon-Visibility-Report/blob/6d5da0e/Hilal%20Obs.py#L101-L108) I put this together:

```python
import matplotlib.pyplot as plt
import math
import numpy
from skyfield.api import Topos, load

ts = load.timescale()
t = ts.utc(2022, 4, 30, 20, 27, 41) # timestamp of the map above

planets = load('de405.bsp')
earth = planets['earth']
sun = planets['sun']
moon = planets['moon']

moon_earth_distance = (moon.at(t) - earth.at(t)).distance().km
sun_earth_distance = (sun.at(t) - earth.at(t)).distance().km

def WidthMoon(AltO, AziO, AltS, AziS, parallax):
    # Yallop 1998, page 3
    GeoAltO = AltO + parallax
    return 0.27245 * parallax * (1 + math.sin(math.radians(GeoAltO)) * math.sin(math.radians(parallax))) * (1 - math.cos(math.radians(AltS - GeoAltO)) * math.cos(math.radians(AziS - AziO)))

H = numpy.ndarray(shape=(18, 36))
for lat in range(0, 180, 10):
    for lng in range(0, 360, 10):
        observer = (earth + Topos(latitude_degrees=90 - lat, longitude_degrees=lng - 180, elevation_m=0)).at(t)
        sun_alt, sun_az, sun_distance = observer.observe(sun).apparent().altaz()
        moon_alt, moon_az, moon_distance = observer.observe(moon).apparent().altaz()
        horizontal_parallax = math.degrees(math.asin(6378.14 / moon_earth_distance)) # is this correct?

        WMoon = WidthMoon(moon_alt.degrees, moon_az.degrees, sun_alt.degrees, sun_az.degrees, horizontal_parallax)
        arcv = moon_alt.degrees - sun_alt.degrees + horizontal_parallax # is this correct?
        
        Wi = WMoon * 60
        H[lat // 10, lng // 10] = (arcv - (11.8371 - 6.3226 * Wi + 0.7319 * Wi * Wi - 0.1018 * Wi * Wi * Wi)) / 10

plt.imshow(H)
plt.show()
```

which gives

![image](https://user-images.githubusercontent.com/833473/167363184-86384514-962d-492b-a41f-05ec8e9a8dcb.png)

and over the days I've tested so many other mix of codes but got more or less same result.

Do you know how such maps can be drawn or does Skyfield provide methods to facilitate drawing of such maps or at least parts of it? Thanks!

**Comment by brandon-rhodes** [comment]
> Do you know how such maps can be drawn or does Skyfield provide methods to facilitate drawing of such maps or at least parts of it? Thanks!

Since no other Skyfield users have weighed in yet, I'll at least offer the answer that: no, alas, Skyfield does not have any built-in facility for drawing a map like that, so you're doing the right thing by trying to come up with code of your own. I'm not familiar with any of the math specific to crescent moon visibility, and at the moment I don't have time to dive into it myself.

Thanks for the link to the moon sighting site, though, those are pretty maps, and I didn't know there was a site that provided sighting data. (Though for some reason my first reaction to the red-letter instructions "Please DO NOT click on maps on the right, otherwise you might get unexpected results" was to immediately and emphatically click on both the maps on the right.)

**Comment by hidp123** [comment]
> Hi! I am trying to create a kind of map which have puzzled me for days which at the end should look like this:
> 
> <img alt="image" width="450" src="https://user-images.githubusercontent.com/833473/167357775-1b59ca98-f462-4bb3-a75d-69b0a316af9d.png">
> 
> [source](https://www.moonsighting.com/visibility.html)
> 
> From [this paper](https://web.archive.org/web/20060628073223/http://web.lums.edu.pk:80/~casm/winterconference-fullpapers/shahidqureshi-paper.pdf), page 13 and 14, I learnt one formulation to create such maps is called Yallop which here is [its paper](https://www.oocities.org/hazarry/astronomy/yallop_1997.pdf) (and from [this code](https://github.com/AminKH/MoonSun/blob/19d2e5e3fe2bee657ce24e306162f55e6989687e/Fortran/MoonSun.f95#L3293-L3297) I learnt there are other formulations also such as Odeh, [its paper](https://bruneiastronomy.org/web/wp-content/uploads/2015/07/New_Criterion_Crescent.pdf), which should give the same but different intensity)
> 
> I searched the term Yallop in GitHub and found [this implementation](https://github.com/aloistr/swisseph/blob/d9476bf5d6aefd85f07c79ccb3c0df8fd0a3e706/swehel.c#L1934) and in combination of [this code from Swiss Ephemeris](https://github.com/msyazwanfaid/Authentication-Algorithm-for-New-Moon-Visibility-Report/blob/6d5da0e/Hilal%20Obs.py#L101-L108) I put this together:
> 
> ```python
> import matplotlib.pyplot as plt
> import math
> import numpy
> from skyfield.api import Topos, load
> 
> ts = load.timescale()
> t = ts.utc(2022, 4, 30, 20, 27, 41) # timestamp of the map above
> 
> planets = load('de405.bsp')
> earth = planets['earth']
> sun = planets['sun']
> moon = planets['moon']
> 
> moon_earth_distance = (moon.at(t) - earth.at(t)).distance().km
> sun_earth_distance = (sun.at(t) - earth.at(t)).distance().km
> 
> def WidthMoon(AltO, AziO, AltS, AziS, parallax):
>     # Yallop 1998, page 3
>     GeoAltO = AltO + parallax
>     return 0.27245 * parallax * (1 + math.sin(math.radians(GeoAltO)) * math.sin(math.radians(parallax))) * (1 - math.cos(math.radians(AltS - GeoAltO)) * math.cos(math.radians(AziS - AziO)))
> 
> H = numpy.ndarray(shape=(18, 36))
> for lat in range(0, 180, 10):
>     for lng in range(0, 360, 10):
>         observer = (earth + Topos(latitude_degrees=90 - lat, longitude_degrees=lng - 180, elevation_m=0)).at(t)
>         sun_alt, sun_az, sun_distance = observer.observe(sun).apparent().altaz()
>         moon_alt, moon_az, moon_distance = observer.observe(moon).apparent().altaz()
>         horizontal_parallax = math.degrees(math.asin(6378.14 / moon_earth_distance)) # is this correct?
> 
>         WMoon = WidthMoon(moon_alt.degrees, moon_az.degrees, sun_alt.degrees, sun_az.degrees, horizontal_parallax)
>         arcv = moon_alt.degrees - sun_alt.degrees + horizontal_parallax # is this correct?
>         
>         Wi = WMoon * 60
>         H[lat // 10, lng // 10] = (arcv - (11.8371 - 6.3226 * Wi + 0.7319 * Wi * Wi - 0.1018 * Wi * Wi * Wi)) / 10
> 
> plt.imshow(H)
> plt.show()
> ```
> 
> which gives
> 
> ![image](https://user-images.githubusercontent.com/833473/167363184-86384514-962d-492b-a41f-05ec8e9a8dcb.png)
> 
> and over the days I've tested so many other mix of codes but got more or less same result.
> 
> Do you know how such maps can be drawn or does Skyfield provide methods to facilitate drawing of such maps or at least parts of it? Thanks!

Did you manage to make some progress? I'm looking to do the same.

**Reply by ebraminio** [reply]
I gave up at some point but am guessing my approach was wrong as most likely I should first calculate sunset time and use the time for each point but not sure.

**Reply by hidp123** [reply]
You need to calculate the q value for each location at it's "best time". Then give the q value a code (A till F - see Yallop's paper), then plot that colour for each location. Looks like you are missing the "best time" calculations for each location. "Best time" is 5/9th the time between sunset and moonset according to Yallop.

**Reply by ebraminio** [reply]
> Looks like you are missing the "best time" calculations for each location.

True, I suspected that also as said

> "Best time" is 5/9th the time between sunset and moonset according to Yallop.

I didn't know this part, definitely useful.

**Reply by hidp123** [reply]
https://astro.ukho.gov.uk/moonwatch/background.html
https://astro.ukho.gov.uk/download/NAOTN69.pdf
https://www.astronomycenter.net/pdf/yallop_1997.pdf (searchable pdf)

**Comment by hidp123** [comment]
**Update:**
See here for progress made on this topic: [crescent-moon-visibility](https://github.com/crescent-moon-visibility/crescent-moon-visibility)
